### ISSUE TITLE
Character ranges checks could be more strict, and they are not portable

### ISSUE THREAD

**Issue by nicolasrochette** [koalaman/shellcheck#985]
#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

-----

Hi. This is the first time I post here and I hope that I have not missed something while doing the above checks on the latest version, or misunderstood this issue.

I spent time this morning debugging a Bash expression of the type `[[ "$var" =~ ^[a-zA-Z_-.]+$ ]]`. There are two issues with the `[^a-zA-Z_-.]` regex pattern.

First, I obviously meant `_`, `-`, or `.`, not the range between `_` and `.`. The pattern should be `[a-zA-Z_.-]` with `-` at the end.

Second, character ranges may depend on the locale. When `$LC_COLLATE` is not set to `C` (e.g. in my case it was undefined and `$LANG` was `en_US.UTF-8`) the behavior of ranges is counter-intuitive. (I note that, on the right of Bash's `=~` operator, `a-z` and `A-Z` specifically seem to keep a consistent meaning.)

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
#shellcheck disable=SC2050 # This expression is constant. Did you forget the $ on a variable?

touch a z A Z

export LC_COLLATE=C
ls [a-z]
[[ "ABC\\" =~ ^[a-z.-_]+$ ]] && echo "1a matches" # case 1a
[[ "abc" =~ ^[a-z_-.]+$ ]] && echo "2a matches" # case 2a

export LC_COLLATE=en_US.UTF-8
ls [a-z]
[[ "abc" =~ ^[a-z_-.]+$ ]] && echo "2b matches" # case 2b
[[ "ABC\\" =~ ^[a-z.-_]+$ ]] && echo "1b matches" # case 1b
[[ "ABC\\" =~ ^[a-z_-.]+$ ]] && echo "1c matches" # case 1c
```
Output:
```
a  z
1a matches
a  A  z
2b matches
```

Case 1a matches because characters `ABC\` are between `.` (ascii 46) and `_` (ascii 95).

Case 2a does not because `_-.` (95-46) is not a valid range (Bash 4.1/4.3 do not issue warnings).

The second `ls` displays file `A` because the ordering underlying the range is `aAbB...zZ`.

Case 2b matches because `.` comes after `_` in this locale.

And cases 1b and 1c do not match respectively because `.-_` isn't a valid range anymore, and `_-.` is a quite limited range.

#### Here's what I wanted or expected to see:

1. `_-.`: To include a literal `-`, make it the first or last character

2. Bash character ranges are not portable, prefer predefined classes such as [:lower:]
